File: travel_buddy_mobile/lib/services/config_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class ConfigService {
  static const String baseUrl = 'https://travelbuddy-backend-h5hqhqhqhqhqhqhq.azurewebsites.net';
  
  static Future<Map<String, dynamic>> getFirebaseConfig() async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/api/config/firebase'),
        headers: {'Content-Type': 'application/json'},
      ).timeout(Duration(seconds: 10));
      
      if (response.statusCode == 200) {
        return json.decode(response.body);
      }
      throw Exception('HTTP ${response.statusCode}: ${response.body}');
    } catch (e) {
      print('ConfigService error: $e');
      // Fallback to local config if backend fails
      return {
        'apiKey': 'fallback-key',
        'authDomain': 'travelbuddy-2d1c5.firebaseapp.com',
        'projectId': 'travelbuddy-2d1c5',
        'storageBucket': 'travelbuddy-2d1c5.firebasestorage.app',
        'messagingSenderId': '45425409967',
        'appId': '1:45425409967:android:8808750a02d7c77356b95a'
      };
    }
  }
}

File: frontend/src/services/configService.ts
export interface RuntimeConfig {
  apiBaseUrl: string;
  firebase: {
    apiKey: string;
    authDomain: string;
    projectId: string;
    storageBucket: string;
    messagingSenderId: string;
    appId: string;
  };
  googleMapsApiKey: string;
  unsplash: {
    accessKey: string;
  };
}

class ConfigService {
  private config: RuntimeConfig | null = null;
  private configPromise: Promise<RuntimeConfig> | null = null;

  async getConfig(): Promise<RuntimeConfig> {
    if (this.config) {
      return this.config;
    }

    // Skip API call, use build-time config directly
    this.config = this.getBuildTimeConfig();
    return this.config;
  }

  private async fetchConfig(): Promise<RuntimeConfig> {
    const baseUrl = import.meta.env.PROD
      ? window.location.origin
      : (import.meta.env.VITE_API_BASE_URL || 'https://travelbuddy-b2c6hgbbgeh4esdh.eastus2-01.azurewebsites.net');

    try {
      const response = await fetch(`${baseUrl}/api/runtime-config`);
      if (!response.ok) {
        throw new Error(`Config fetch failed: ${response.status}`);
      }
      const runtimeConfig = await response.json();

      return {
        apiBaseUrl: baseUrl,
        firebase: runtimeConfig.firebase,
        googleMapsApiKey: import.meta.env.VITE_GOOGLE_MAPS_API_KEY || '',
        unsplash: {
          accessKey: import.meta.env.VITE_UNSPLASH_ACCESS_KEY || ''
        }
      };
    } catch (error) {
      console.error('Failed to fetch runtime config, using fallback:', error);
      return this.getBuildTimeConfig();
    }
  }

  private getBuildTimeConfig(): RuntimeConfig {
    return {
      apiBaseUrl: import.meta.env.PROD
        ? window.location.origin
        : (import.meta.env.VITE_API_BASE_URL || 'http://localhost:3000'),
      firebase: {
        apiKey: import.meta.env.VITE_FIREBASE_API_KEY || '',
        authDomain: import.meta.env.VITE_FIREBASE_AUTH_DOMAIN || '',
        projectId: import.meta.env.VITE_FIREBASE_PROJECT_ID || '',
        storageBucket: import.meta.env.VITE_FIREBASE_STORAGE_BUCKET || '',
        messagingSenderId: import.meta.env.VITE_FIREBASE_MESSAGING_SENDER_ID || '',
        appId: import.meta.env.VITE_FIREBASE_APP_ID || ''
      },
      googleMapsApiKey: import.meta.env.VITE_GOOGLE_MAPS_API_KEY || '',
      unsplash: {
        accessKey: import.meta.env.VITE_UNSPLASH_ACCESS_KEY || ''
      }
    };
  }


}

export const configService = new ConfigService();
